File: src/jastudio/jastudio/ui/web/sentence/text_analysis_viewmodel.py
from __future__ import annotations

from typing import TYPE_CHECKING

from autoslot import Slots  # pyright: ignore[reportMissingTypeStubs]
from jastudio.ui.web.sentence.candidate_word_variant_viewmodel import CandidateWordVariantViewModel

if TYPE_CHECKING:
    from jastudio.language_services.janome_ex.word_extraction.text_analysis import TextAnalysis
    from jastudio.ui.web.sentence.match_viewmodel import MatchViewModel
    from typed_linq_collections.collections.q_list import QList


class TextAnalysisViewModel(Slots):
    def __init__(self, text_analysis: TextAnalysis) -> None:
        self.analysis: TextAnalysis = text_analysis
        self.candidate_words: QList[CandidateWordVariantViewModel] = text_analysis.indexing_word_variants.select(CandidateWordVariantViewModel).to_list()
        variant_view_models: QList[CandidateWordVariantViewModel] = self.candidate_words
        matches: QList[MatchViewModel] = variant_view_models.select_many(lambda variant_vm: variant_vm.matches).to_list()
        self.displayed_matches:QList[MatchViewModel] = matches.where(lambda match: match.is_displayed).to_list()

File: src/src_dotnet/JAStudio.Core/UI/Web/Sentence/TextAnalysisViewModel.cs
using System.Collections.Generic;
using System.Linq;
using JAStudio.Core.Configuration;
using JAStudio.Core.LanguageServices.JanomeEx.WordExtraction;
using JAStudio.Core.Note.Collection;

namespace JAStudio.Core.UI.Web.Sentence;

public class TextAnalysisViewModel
{
    public TextAnalysis Analysis { get; }
    public List<CandidateWordVariantViewModel> CandidateWords { get; }
    public List<MatchViewModel> DisplayedMatches { get; }

    public TextAnalysisViewModel(TextAnalysis textAnalysis, Settings settings, VocabCollection vocab)
    {
        Analysis = textAnalysis;
        CandidateWords = textAnalysis.IndexingWordVariants
            .Select(v => new CandidateWordVariantViewModel(v, settings, vocab))
            .ToList();
        
        var matches = CandidateWords.SelectMany(variantVm => variantVm.Matches).ToList();
        DisplayedMatches = matches.Where(match => match.IsDisplayed).ToList();
    }
}
